### PR TITLE
Add support for SomfyHeatingTemperatureInterface

### PR DESCRIPTION
**Pull request by egguy**
Updated implementation of the SomfyHeatingTemperatureInterface.

- Use the temperature of the thermostat
- Added 2 NumberSettings, but wasn't able to see them.
- Add "Impossible to change from the box" mode selection (cooling/heating).
- Mostly working
- Need to add some constant to pyOverKiz

Sorry if the pull request is not in the good format. It's mostly my first one.

<a href="https://gitpod.io/#https://github.com/iMicknl/ha-tahoma/pull/703"></a>

### CODE REVIEW

**Comment by tetienne** [comment]
Hi, many thanks for this 1st contribution, any help are welcome 🎉  I haven’t yet look at your code, as you can see you have too many files modified. It’s probably because your branch is out of sync with our repo. To solve this:

```bash
git fetch --all
git rebase upstream/master
# Check you haven’t any conflicts, if yes solve them. 
git push --force-with-lease
```

Once done, your PR will be ready for our review :)

**Comment by egguy** [comment]
Ok, I followed your procedure, I hope I didn't make a mistake.

**Comment by tetienne** [comment]
I’ve updated the base branch. Apparently, we only have your modification now.

File: custom_components/tahoma/climate_devices/valve_heating_temperature_interface.py
Comment (iMicknl) [inline custom_components/tahoma/climate_devices/valve_heating_temperature_interface.py:1]: I think you accidentally committed this one, or started at a feature branch. Can you remove this file and push again? :)

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (iMicknl) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:1]:
```suggestion
"""Support for Somfy Heating Temperature Interface."""
```

File: custom_components/tahoma/climate_devices/valve_heating_temperature_interface.py
Comment (egguy) [reply custom_components/tahoma/climate_devices/valve_heating_temperature_interface.py]: I continued from the feature branch, SomfyHeatingTemperatureInterface. That can explain it.

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (iMicknl) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:59]: We need to add them to `OverkizCommandParam` indeed, as you mentioned. For now this is fine :).

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (iMicknl) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:54]:
```suggestion
    OverkizCommandParam.AUTO: HVAC_MODE_AUTO,
    OverkizCommandParam.MANU: HVAC_MODE_HEAT_COOL,
```

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (iMicknl) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:47]:
```suggestion
    OverkizCommandParam.SECURED: PRESET_AWAY,
    OverkizCommandParam.ECO: PRESET_ECO,
    OverkizCommandParam.COMFORT: PRESET_COMFORT,
    OverkizCommandParam.FREE: PRESET_NONE,
```

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (iMicknl) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:151]:
```suggestion
        if OverkizState.OVP_HEATING_TEMPERATURE_INTERFACE_OPERATING_MODE in self.device.states:
            return TAHOMA_TO_HVAC_MODES[
                self.executor.select_state(OverkizState.OVP_HEATING_TEMPERATURE_INTERFACE_OPERATING_MODE)
```

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (iMicknl) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:145]:
```suggestion
        if OverkizState.CORE_ON_OFF in self.device.states:
```

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (iMicknl) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:209]:
```suggestion
                OverkizCommand.SET_COMFORT_TEMPERATURE, temperature
```

**Review by iMicknl — CHANGES_REQUESTED** [review]
Awesome work! I added a few remarks, however it would be great if you can check which enums you can reuse already!

See https://github.com/iMicknl/python-overkiz-api/blob/main/pyoverkiz/enums/command.py and https://github.com/iMicknl/python-overkiz-api/blob/main/pyoverkiz/enums/state.py.

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (egguy) [reply custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:35]: Thanks, forgot to remove it. It's a remain from trying to add the mode change through the climate entity. Moved to a Select

File: custom_components/tahoma/climate_devices/valve_heating_temperature_interface.py
Comment (iMicknl) [reply custom_components/tahoma/climate_devices/valve_heating_temperature_interface.py]: Good chance that I messed up that branch before :D.

**Comment by egguy** [comment]
Thanks for the review, tried to find as much as possible constants from the overkiz library. Seems I've missed quite a few... 😥.

I will try to do all the changes tonight.

**Comment by iMicknl** [comment]
No worries! I will have a better look afterwards, but good to have it as clean as possible to start :).

Since you own this device, this could be a device that we can bring to core more easily if the code already matches the standards.

**Comment by egguy** [comment]
Taken time to update the number inputs to this convention at the same time.

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (iMicknl) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:204]:
```suggestion
            OverkizState.OVP_HEATING_TEMPERATURE_INTERFACE_SETPOINT_MODE
```

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (iMicknl) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:203]: In theory you could here check as well for `self.preset_mode`, since that is using the `OVP_HEATING_TEMPERATURE_INTERFACE_SETPOINT_MODE` as well. This would mean that you don't have to pull the state again, but can use the already retrieved state.

@tetienne / @vlebourl what would be your preference?

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (iMicknl) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:173]: Please add this to https://github.com/iMicknl/python-overkiz-api/pull/310/files

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (iMicknl) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:158]: Please add to https://github.com/iMicknl/python-overkiz-api/pull/310/files

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (iMicknl) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:138]: Not sure if this if statement is required, since `select_state` will return None if the state cannot be found?

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (iMicknl) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:190]: Why is the target temperature `None` when it is in auto mode? Is that how it works with a heater?

**Review by iMicknl — CHANGES_REQUESTED** [review]
Thanks again! Just some minor questions / remarks.

File: custom_components/tahoma/number.py
Comment (iMicknl) [inline custom_components/tahoma/number.py:34]: Should we change this to temperature? They are in the config section already, thus perhaps best to reflect what the setting does?

 ```suggestion
        icon="mdi:thermometer",
```

Or `mdi:home-thermometer-outline` for room temperature and `mdi:sun-thermometer-outline` for Freeze protection temperature.

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (egguy) [reply custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:190]: So this is the interesting part. This is an "ambiance thermostat".
It has 2 working modes:
- Auto:  Run the planned programmed scenario (ex: eco from 23H00 - 06H00 and comfort from 06H00 - 23H00), so it's the HVAC_MODE_AUTO behavior
- Manu: Run the heating with the current setting (ex: manu - eco = 17 deg, manu - comfort : 19 deg)

When you change the temperature setting in manual mode, it changes the temperature used in the auto (ex: manu - eco set at 18 deg, then in auto when it's the eco's time the temperature is set to 18 deg).

Here's the manual of one of the numerous variation (couldn't find an English version) https://damrexelprod.blob.core.windows.net/medias/2fb79f22-2075-4f5a-93d3-112b14a4b8c2)

There's also the cooling mode which reverse how it works (open the valve when over the set temperature, close if under), it's used in heated/cooled floor for exemple.

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (egguy) [reply custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:203]: I think it can be removed.
There's no way to set the current temperature in auto mode, if it's in a manual mode, then you change current preset (eco / comfort / freeze).
That's why I've asked if it's possible to changes the feature dynamically to allow changing temperature only in manual mode.

Comment (egguy) [reply custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:206]: I've done some test, it works well with the SUPPORT_TARGET_TEMPERATURE, but  there's no way to set a derogated temperature in auto mode.

It affects the current mode (eco/comfort) and stay persisted if you change to manual mode. This is contrary to how the physical device work.

I think it should stay disabled, and should be done thought the NumberEntity instead to prevent surprise from the users.

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (tetienne) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:72]: We want to get rid of this pattern. Can you please mimic https://github.com/iMicknl/ha-tahoma/pull/704/files and tell me if it works correctly on your branch?
It will ease a lot the code.

File: custom_components/tahoma/number.py
Comment (tetienne) [inline custom_components/tahoma/number.py:60]: On your device, you can set the 3 temperatures at the same time? The 3 commands are always available or only one?

Comment (iMicknl) [reply custom_components/tahoma/number.py:60]: Could it be because this is set to config? So it will only be visible on the device page via Integrations -> Devices.

File: custom_components/tahoma/select.py
Comment (iMicknl) [inline custom_components/tahoma/select.py:46]:
```suggestion
        options=[OverkizCommandParam.HEATING, OverkizCommandParam.COOLING],
```

File: custom_components/tahoma/number.py
Comment (egguy) [reply custom_components/tahoma/number.py:60]: This is all I have for the device: 
image
We can see the mode change, but not the 3 others number settings.

Comment (iMicknl) [reply custom_components/tahoma/number.py:60]: Do you have any errors in your log and which version of HA do you run? I think I have some feeling what is wrong. If you apply this fix -> https://github.com/iMicknl/ha-tahoma/pull/667, your dev works again.

Comment (egguy) [reply custom_components/tahoma/number.py:60]: I will try to apply the fix. I'm running: Home Assistant 2021.12.9

File: custom_components/tahoma/number.py
Comment (tetienne) [inline custom_components/tahoma/number.py:33]: If we follow what we have done for the other numbers, it must be `Eco Room Temperature`.

@iMicknl what do you think?

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (tetienne) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:70]: So it works well?

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (tetienne) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:121]: If the condition is not true, the method return nothing.

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (tetienne) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:154]:
```suggestion
        _LOGGER.error("Unkown mode: %s", mode)
```

**Review by tetienne — CHANGES_REQUESTED** [review]

File: custom_components/tahoma/number.py
Comment (iMicknl) [reply custom_components/tahoma/number.py:33]: I would like that better I think, or if it doesn't make sense just Eco Temperature (capitalized).

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (egguy) [reply custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:70]: It's working, but linked_device shouldn't be an @property. 

The way to get the temperature also feel unnatural, it's different from the other devices. (no executor in the device), otherwise it's perfectly working.

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (egguy) [reply custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:121]: Should I return None when there's nothing ? From the climate entity documentation, it seems to be the default value.

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (tetienne) [reply custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:70]: Good point for the property, they cannot take an argument.

About your feeling, it's normal. But we want to migrate all the other climate devices to this logic way more simple.

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (tetienne) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:127]: Same comment that https://github.com/iMicknl/ha-tahoma/pull/703/files#r784721031

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (egguy) [reply custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:90]: We have the risk of logging `None` values, which can be normal if the state is not found (shouldn't happen).
With the check, we can quickly discover new mode instead of having noise.

Perhaps put in "Warning" level ?

Comment (tetienne) [reply custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py]: As you said it should not happen. If it's the case, you will log an error.

**Comment by iMicknl** [comment]
@egguy I just updated your branch to use the latest fixes for the coordinator (double casting) and updates pyoverkiz to 1.1.1.

**Comment by egguy** [comment]
Pulled the changes. Now the OverkizNumber input are working. (need to comment the step parameter)
What's remain to allow the pull request to be integrated ? 

Ps: What is the relation between ha-tahoma and overkiz in core ? Is ha-tahoma the test integration ?

File: custom_components/tahoma/select.py
Comment (iMicknl) [inline custom_components/tahoma/select.py:48]: Did you add setOperatingMode to pyoverkiz already?

**Review by iMicknl — DISMISSED** [review]
Looks good to me. Let's merge if another maintainer approves :-).

**Comment by iMicknl** [comment]
> Ps: What is the relation between ha-tahoma and overkiz in core ? Is ha-tahoma the test integration ?

This integration is a bit behind on some parts now, since we had great code reviews in core. However, the aim is to migrate the integration 1:1 to core, but that will take quite some time. Eventually ha-tahoma will be the test integration, and core will be the production.

See https://github.com/iMicknl/ha-tahoma/issues/697 for the progress and what is still due (a lot! 😄 ).

File: custom_components/tahoma/number.py
Comment (iMicknl) [inline custom_components/tahoma/number.py:92]: Maybe we should try to keep this out of this PR. We have https://github.com/iMicknl/ha-tahoma/pull/667 ready for when 2022.2 drops.

File: custom_components/tahoma/select.py
Comment (egguy) [reply custom_components/tahoma/select.py:48]: Yes, will push the change.

File: custom_components/tahoma/number.py
Comment (egguy) [reply custom_components/tahoma/number.py:92]: Will the step parameter be working  in the 2022.2  ?

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (iMicknl) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:140]: Perhaps add some comments in this method? This one is a bit unclear if you don't own the device and don't understand why target temperature is different / why target temperature can be empty.

**Comment by egguy** [comment]
> > Ps: What is the relation between ha-tahoma and overkiz in core ? Is ha-tahoma the test integration ?
> 
> This integration is a bit behind on some parts now, since we had great code reviews in core. However, the aim is to migrate the integration 1:1 to core, but that will take quite some time. Eventually ha-tahoma will be the test integration, and core will be the production.
> 
> See #697 for the progress and what is still due (a lot! smile ).

The overkiz integration is mostly for covers at the moment ? Or does it manage the other types of devices ?

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (egguy) [reply custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:140]: I think it can also be removed. It's more logical to configure through the external number than target_temperature.

Using this function, reconfigure the current mode, even in auto mode, which can be surprising. I need to see if I can set the override temperature in auto mode.

**Comment by iMicknl** [comment]
> The overkiz integration is mostly for covers at the moment ? Or does it manage the other types of devices ?

The Overkiz integration (1:1 copy from ha-tahoma) doesn't have the cover platform yet. See the issue I linked to see which platforms are already merged.

File: custom_components/tahoma/number.py
Comment (iMicknl) [reply custom_components/tahoma/number.py:92]: Yes :)

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (egguy) [reply custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:152]: Done

**Comment by vlebourl** [comment]
Are @tetienne 's requested changes included? if so, we should dismiss the review and proceed.

**Comment by egguy** [comment]
I have applied the requested changes. For the codespell error, it comes from a URL. :s

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (iMicknl) [inline custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:71]: Let's see if we can add a codespell ignore somewhere here. Couldn't find the required # format yet on mobile.

File: custom_components/tahoma/select.py
Comment (iMicknl) [inline custom_components/tahoma/select.py:44]:
```suggestion
        name="Operating Mode",
```

File: custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py
Comment (egguy) [reply custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py:71]: Or i can just remove the link.

Comment (iMicknl) [reply custom_components/tahoma/climate_devices/somfy_heating_temperature_interface.py]: Fine by me as well :)

**Comment by egguy** [comment]
I have added the pre-commit to my env to prevent more problems in the future.

**Review by iMicknl — APPROVED** [review]
Thanks again! Happy to merge if the other users confirms that this works. https://github.com/iMicknl/ha-tahoma/issues/705

**Comment by iMicknl** [comment]
@egguy do you perhaps want to port this one to core? :)
Otherwise I will have a look in the coming weeks.

**Comment by egguy** [comment]
Hi, i'm already doing the port ATM. Hope to be able to commit soon.

**Pull request merged**
